### PR TITLE
Extend the wwan connection tests to support multiple cycle connection tests (New)

### PR DESCRIPTION
**Pull request by stanley31huang**
## Description
Extend the wwan connection tests to support multiple cycle connection tests, we found an issue that the LTE connection could not be established at second round.

As discussed internally with the QA team, we have all agreed that the default number of test runs for WWAN connections tests is 2.

## Resolved issues
Here is the support ticket: 00379830.

Our SWE has identified this issue is related to the modem firmware and modem-manager snap, and they are trying to add a patch to modem-manager snap as a workaround.

## Documentation
N/A

## Tests
https://certification.canonical.com/hardware/202302-31253/submission/386810/test-results/

### CODE REVIEW

**Review by pieqq — CHANGES_REQUESTED** [review]
Thanks for this contribution! This is a good idea to try to connect WWAN multiple times, this will for sure help with regression testing.

A few things:

# Default number of iterations

The default number of iterations is something that should be discussed with other QA teams, especially PC QA. With your proposal, if the `WWAN_CONNECTION_TEST_COUNT` environment variable is not set, there will only be one connection test. Maybe PC QA and other IoT QA projects would benefit from having several connection attempts?

# What if connection fails at first attempt?

Should we still try `X` times, adding `X` failures to the test report?

# Splitting connection jobs Vs. connecting multiple times in the same job

Is it important to have multiple jobs (resulting in several lines in the test report), or could the number of iterations provided to the `wwan_tests.py` script, and the connection would be attempted `X` times instead?

# Code duplication

I think you could combine the new `wwan_resource_with_iterations.py` with the existing `wwan_tests.py`. The latter already provides the functionality to print out the resource; you could just add the `--iterations` argument to the `wwan_tests.py` script and use the same mechanism you developed in `wwan_resource_with_iterations.py` (repeat the resource info for each iteration).

By doing this, you

- prevent code duplication
- reduce the number of scripts to maintain
- reduce the number of additional tests

Since you only want the connection test to be run multiple times, and not the other tests based on `wwan_resource`, an easy fix is to add

```
template-filter: wwan_resource.iteration == '1'
```

to these jobs. They will only be run for the first iteration.

I've tried it locally and it seems to work; here are the modifications I made to the WWAN jobs using `wwan_resource`:

```diff
diff --git a/providers/base/units/wwan/jobs.pxu b/providers/base/units/wwan/jobs.pxu
index 99cf4fb45..07d9fe901 100644
--- a/providers/base/units/wwan/jobs.pxu
+++ b/providers/base/units/wwan/jobs.pxu
@@ -27,7 +27,7 @@ requires:
   snap.name == 'modem-manager' or package.name == 'modemmanager'
 
 unit: template
-template-resource: wwan_resource_with_iteration
+template-resource: wwan_resource
 template-unit: job
 id: wwan/gsm-connection-{manufacturer}-{model}-{hw_id}-cycle{iteration}-auto
 template-id: wwan/gsm-connection-manufacturer-model-hw_id-auto
@@ -59,6 +59,7 @@ requires:
 unit: template
 template-resource: wwan_resource
 template-unit: job
+template-filter: wwan_resource.iteration == '1'
 id: wwan/check-sim-present-{manufacturer}-{model}-{hw_id}-auto
 template-id: wwan/check-sim-present-manufacturer-model-hw_id-auto
 _summary:  Check if a SIM card is present in a slot connected to the modem
@@ -79,6 +80,7 @@ requires:
 unit: template
 template-resource: wwan_resource
 template-unit: job
+template-filter: wwan_resource.iteration == '1'
 id: wwan/verify-sim-info-{manufacturer}-{model}-{hw_id}
 template-id: wwan/verify-sim-info-manufacturer-model-hw_id
 depends: wwan/check-sim-present-{manufacturer}-{model}-{hw_id}-auto
```

**Comment by stanley31huang** [comment]
@pieqq 
Thanks for reviewing this PR and the suggestions. I like the idea to integrate the iteration into the `wwan_resource` job, to avoid the code duplication.

For the connections fail at the first attempt, I think it still worth to launch the rest of the connection tests, because of we could understand the failure rate is 100% or not. But it would be better to add a dependency in this job which is make sure the SIM card has been detected.

For the `Splitting connection jobs Vs. connecting multiple times in the same job`, I personally prefer the way to `splitting connections jobs`, because of it's easily to know the overall test results, no need to go through the details from the output of job.

File: providers/base/units/wwan/resource.pxu [hunk 1]
```diff
@@ -15,3 +15,15 @@ command: wwan_tests.py resources
...
 flags: preserve-locale
+
+unit: job
+id: wwan_resource_with_iteration
+category_id: wwan
+plugin: resource
+_summary: Gather device info about WWAN modems
+_description: Gather device info about WWAN modems
+command: wwan_resource_with_iterations.py -i "${WWAN_CONNECTION_TEST_COUNT:-1}"
+environ: WWAN_CONNECTION_TEST_COUNT
```
[the remark this answers is not in the document]
Comment (stanley31huang) [reply providers/base/units/wwan/resource.pxu:26]: will check with QA internally, thanks.

**Comment by stanley31huang** [comment]
Converted this PR to draft, due to I am going to re-base this branch after another PR has been merged.
https://github.com/canonical/checkbox/pull/1323

File: providers/base/units/wwan/resource.pxu (hunk 1 above)
[the remark this answers is not in the document]
Comment (stanley31huang) [reply providers/base/units/wwan/resource.pxu]: As discussed internally with the QA team ( including members who work for PC projects), we have all agreed that the default number of test runs for WWAN connections tests is 2.

**Comment by pieqq** [comment]
Following today's Office Hours meeting, we discussed with QA and Checkbox team regarding this PR.

If the goal is "just" to run the jobs twice, then maybe we could just use a `sibling`, and add that new job to the test plan.

**Comment by stanley31huang** [comment]
close this one, due to the changes are bit messy. Use https://github.com/canonical/checkbox/pull/1534 instead.

**Pull request closed without merging**